### PR TITLE
port: Fix: add cache for luis recognizer result to avoid extra requests

### PR DESCRIPTION
**Pull request by cosmicshuai**
closes: #3806 
In order to avoid two LUIS network calls from the child dialog for the same user query when interruptions are enabled.
In this PR, Luis Recognizer would cache the luis result for each turn, using endpoint combines with application id as the key.

### CODE REVIEW

**Comment by cosmicshuai** [comment]
@joshgummersall 
Hi Josh, as [Steve mentioned](https://github.com/microsoft/botbuilder-dotnet/issues/5593#issuecomment-861088870), the LuisRecognizer already has the cache. In order to port the test, I've changed a lot code in testScript and LuisAdaptiveRecognizer. I am wondering whether it is worth to do that? Or is there a better way to write this test?

File: libraries/botbuilder-core/src/middlewareSet.ts [hunk 1]
```diff
@@ -77,6 +78,15 @@ export class MiddlewareSet implements Middleware {
         this.use(...middlewares);
     }
 
+    [Symbol.iterator]() {
```
Comment (joshgummersall) [inline libraries/botbuilder-core/src/middlewareSet.ts:81]: Why is this necessary? I don't see any iteration through middlewares?

File: libraries/botbuilder-core/src/middlewareSet.ts [hunk 2]
```diff
@@ -77,6 +78,15 @@ export class MiddlewareSet implements Middleware {
...
+        return {
+            next(): IteratorResult<MiddlewareHandler> {
+                this.count++;
```
Comment (joshgummersall) [inline libraries/botbuilder-core/src/middlewareSet.ts:84]: If we were to keep this iterator, you would actually want to do:

```typescript
[Symbol.iterator]() {
  const count = 0;
  return {
    return { done: count >= this.middleware.length, value: this.middleware[count++] };
  }
}
```

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol

File: libraries/botbuilder-core/src/botAdapter.ts [hunk 3]
```diff
@@ -28,7 +28,7 @@ import { TurnContext } from './turnContext';
  * [Middleware](https://docs.microsoft.com/azure/bot-service/bot-builder-concept-middleware).
  */
 export abstract class BotAdapter {
-    protected middleware: MiddlewareSet = new MiddlewareSet();
+    public middleware: MiddlewareSet = new MiddlewareSet();
```
Comment (joshgummersall) [inline libraries/botbuilder-core/src/botAdapter.ts:31]: Why is this necessary?

File: libraries/botbuilder-dialogs-adaptive-testing/tests/luisAdaptiveRecognizer.test.js [hunk 4]
```diff
@@ -1,6 +1,6 @@
 const path = require('path');
 const { LuisAdaptiveRecognizer, LuisBotComponent } = require('botbuilder-ai');
-const { MockLuisLoader, MockLuisRecognizer, TestUtils, useMockLuisSettings } = require('../lib');
+const { MockLuisLoader, MockLuisRecognizer, TestUtils, useMockLuisSettings, AssertTelemetryContains } = require('../lib');
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive-testing/tests/luisAdaptiveRecognizer.test.js:3]: Where is `AssertTelemetryContains` used?

File: libraries/botbuilder-core/src/middlewareSet.ts (hunk 1 above)
Comment (cosmicshuai) [reply libraries/botbuilder-core/src/middlewareSet.ts:81]: Yes, it is not necessary. The first implementation was iterating through middlewareset to get TelemetrtyMiddleWare. Now, it is changed to set to turnState.

File: libraries/botbuilder-core/src/botAdapter.ts (hunk 3 above)
Comment (cosmicshuai) [reply libraries/botbuilder-core/src/botAdapter.ts:31]: the same reason as before

File: libraries/botbuilder-dialogs-adaptive-testing/src/mocks/mockLuisRecognizer.ts [hunk 5]
```diff
@@ -25,6 +25,7 @@ import {
     MockHttpRequestMiddlewareKey,
 } from './mockHttpRequestMiddleware';
 import { HttpResponseMessage } from '../httpRequestMocks/httpResponseMock';
+import { TestTelemetryClient } from '../testTelemetryClient';
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive-testing/src/mocks/mockLuisRecognizer.ts:28]: Is this import unused?

Comment (cosmicshuai) [reply libraries/botbuilder-dialogs-adaptive-testing/src/mocks/mockLuisRecognizer.ts]: removed

**Comment by joshgummersall** [comment]
Nice work! Is this "Ready for Review" aka ready to merge?

**Review by joshgummersall — APPROVED** [review]
